Add support for --hex-blob option in mysqldump

vdechef opened the pull request.

dumblob commented on mysql2sqlite:
```diff
@@ -121,16 +121,53 @@ inView != 0 { next }
-    printerr( \
-      NR ": WARN Hex number trimmed (length longer than 16 chars)." )
+  # get content of VALUES statement
+  if( match( $0, /VALUES \((.*)\);$/ ) ) {
```
Please change `VALUES` to `(VALUES|values)` (e.g. because of dumps made on Windows). Also try to always add `+` to blank characters as apps have different rules for paddings which could cause problems in some cases. Another thing is backslashes, which I would rather avoid as they have generally a bit convolved rules (at least in the POSIX spec), so `[]` is always the safe and best option (though in this case backslashes are correct :wink:).

So the result might look like `/(VALUES|values) +[(].*[)];$/`.

dumblob commented on mysql2sqlite:
```diff
@@ -121,16 +121,53 @@ inView != 0 { next }
+  # get content of VALUES statement
+  if( match( $0, /VALUES \((.*)\);$/ ) ) {
+    valuesStart = RSTART
+    remaining = substr($0, RSTART+8, RLENGTH-10 )
```
Because we added `+` above, we can't use `+8` here, but need to match it again :cry:.

vdechef commented:
> 
> 
> Thanks for the PR. It look good to me. Please try to address my comments or discuss them and then I'll merge it 😉.

OK, I will fix this later this week. I will also try to improve unit tests.

dumblob commented:
@vdechef take your time (mine is also a "scarcity").

vdechef commented:
I think I'm done.  
* I updated the code with your remarks, and added the comments you asked for
* I also fixed a bug on ENUM regex, and cleaned some \ escaping as you told me to
* I added some unit tests, in a way that it should be easy to extend (this can be discussed actually)

dumblob commented on unit_tests.sh:
```diff
@@ -2,19 +2,153 @@
 
 M2S=./mysql2sqlite
 S=sqlite3
+X=xxd
```
Please use [`od`](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/od.html ) (it's POSIX unlike `xxd`).

dumblob replied:
And when already at it, please add `readonly ` before all these constants :wink:.

vdechef replied:
Ok for the readonly part.

But regarding `xxd`, I cannot replace it by `od` : the purpose of `xxd` usage is to convert hexstring to bytes (`od` does the invert operation). I have just tried to implement a POSIX compliant conversion of hexstring to binary file but I couldn't achieve this. 

I am not very comfortable with shell script so I won't be able to solve this thing without a lot of personal work. If you have a suggestion of solution, I would be glad to test it (and to learn something at the same time ;)

dumblob replied:
>But regarding `xxd`, I cannot replace it by `od`: the purpose of `xxd` usage is to convert hexstring to bytes (`od` does the invert operation).

Oh, sorry, my bad - I didn't look at the usage and blindly wrote about `od`. For the other way round printf with eval should do:
```
eval "printf '$(printf '\\%o\\%o\\%o\\%o' 0x61 0x62 0x63 0x64)'"
```
To convert hex input to separate arguments for printf many ways can be used, but I'd probably again use eval (with `sed`) as it's built-in and therefore significantly faster in parsing then writing a parser using POSIX shell and utilities (like `awk`) manually. To make it more readable, I'd of course save it into a variable instead of going for eval nesting and thus quoting hell on one line :wink:.

dumblob replied:
Hm, thinking about it further, we might replace the inner `printf` with [`bc`](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/bc.html ) as it should allow us to avoid the `eval` without much burden and be even faster than using the builtin shell parsing :wink:. Feel free whatever you feel more comfortable with :wink:.

With `bc` it could look similar to:
```sh
bc <<END
obase=8; ibase=16; 61;62;63;64
END
```

vdechef replied:
Ok, I will try with bc (I wasn't sur that this was posix compliant)

vdechef replied:
done with bc and heredoc

dumblob commented on unit_tests.sh:
```diff
@@ -2,19 +2,153 @@
+H=md5sum
+B64=base64
 
 type "$S" >/dev/null 2>&1 || {
```
Let's squeeze all these into a simple function applied to each "command" to make things terser and more readable :wink:.

vdechef replied:
sure. done.

dumblob commented on unit_tests.sh:
```diff
@@ -2,19 +2,153 @@
+query() {
+  QUERY="$1"
+  DB="$2"
+  RESULT=$(echo "$QUERY" | sqlite3 "$DB" 2>&1)
```
Please use double quotes around `$()` and always use `printf` instead of `echo`. Or if you care about performance, heredoc is significantly more performant than a pipe:
~~~
r="$(sqlite3 "$DB" 2>&1 <<END
$QUERY
END
printf '_')"
~~~

Another and in this case actually more important thing is, that `$()` truncates trailing newline(s). Therefore we should always print one more non-newline character at the very end of `$()` where we care about each byte. I.e. `printf '_'` as I wrote above.

vdechef replied:
I fixed the bugs (nice catch for trailing newlines truncation). 

Regarding heredoc, the problem is with script formatting : it breaks the indentation, so for now - if you don't mind - I won't use it. As we are talking about unit tests, performances are not really important.

vdechef replied:
Finally with `<<-END` and tabs I was able to format it properly.

dumblob commented on unit_tests.sh:
```diff
@@ -2,19 +2,153 @@
+#     2. the .sqlite file to use as database for the query
+#
+query() {
+  QUERY="$1"
```
Please add `[ $# -eq 2 ] || printf 'USAGE: query <query> <dbfilename>\n'`.

vdechef replied:
done

dumblob commented on unit_tests.sh:
```diff
@@ -2,19 +2,153 @@
+  QUERY="$1"
+  DB="$2"
+  RESULT=$(echo "$QUERY" | sqlite3 "$DB" 2>&1)
+  printf "$RESULT"
```
Please never use `printf` like this, it's very dangerous. Use `printf '%s\n' "$RESULT"` instead.

vdechef replied:
done

dumblob commented on unit_tests.sh:
```diff
@@ -2,19 +2,153 @@
+#     2. the expected result of the query (as string)
+#     3. the .sqlite file to use as database for the query
+#
+assert_query() {
```
Please apply here same changes as to the function above.

vdechef replied:
done

dumblob commented on unit_tests.sh:
```diff
@@ -2,19 +2,153 @@
+# NB: dump was produced using `sudo mysqldump --skip-extended-insert --hex-blob --compact --single-transaction testtypes > unit_tests/dump.sql`
+#     To add tests in this dump, just import it into your MySQL instance and re-run the command above.
+UT=./unit_tests
+OUT_script=$UT/test_dump.sqlite
```
Please use double quotes everywhere. Also I'd prefer consistent upper/lowercase variable names (I'm though indifferent to whether you'll use uppercase or lowercase names; I myself use lowercase usually :wink:).

vdechef replied:
Ok for double quotes.

I changed case : I used uppercase for readonly variables, and lowercase for other variables.

dumblob commented on unit_tests.sh:
```diff
@@ -2,19 +2,153 @@
+UT=./unit_tests
+OUT_script=$UT/test_dump.sqlite
+OUT_database=$UT/test_db.sqlite
+rm $OUT_script 2> /dev/null
```
Just for safety, always use `--` before passing file names to any utility (here `rm -- "$OUT_script"). Accidents happen and nobody wants to do some harm :wink:.

vdechef replied:
Indeed

dumblob commented on unit_tests.sh:
```diff
@@ -2,19 +2,153 @@
+rm $OUT_database 2> /dev/null
+$M2S $UT/dump.sql > $OUT_script
+# create sqlite database using generated sqlite script
+cat $OUT_script | $S $OUT_database
```
Please avoid unnecessary `cat` by using plain redirection `"$S" "$OUT_database" < "$OUT_script"` to save pipe (i.e. forking).

vdechef replied:
done

dumblob commented on unit_tests.sh:
```diff
@@ -2,19 +2,153 @@
+OUT_database=$UT/test_db.sqlite
+rm $OUT_script 2> /dev/null
+rm $OUT_database 2> /dev/null
+$M2S $UT/dump.sql > $OUT_script
```
I'd prefer using heredoc here for input to avoid the external `dump.sql` file. That's also why all other testing dumps are in heredocs, though currently unused (waiting for me/someone to finally implement their checking).

vdechef replied:
The idea here was to have a sql script that could be imported into a real database (for debug purpose). This way it is very easy to just add statements in the dump.sql each time a user send you a new error case. You won't even have to write a new test in the sh script to check if the statement breaks mysq2sqlite : you add the statement in dump.sql, run the unit tests, check the resulting sqlite script and sqlite database (if any), fix the awk script, and then if you need to you can add test cases by sending queries to the sqlite database.

IMO if you write into the test script all your statements, it will be harder to maintain.

dumblob replied:
>The idea here was to have a sql script that could be imported into a real database (for debug purpose).

Yep, that's how I do it **before** I make a unit test out of it. After you make a unit test out of it, the idea is though quite the opposite - to **not touch the unit tests at all** and adjust the `myqsl2sqlite` script to match the unit tests expectations. So I actually don't think you would need to adjust/manually_import the unit tests SQL codes any time (and if so, then only due to very major totally breaking changes in the repo and the conversion script itself in which case it doesn't matter what the old "architecture" or "format" was).

Regarding "one huge sql code" serving as a universal test code for every case I initially didn't want to do it because I wanted each test have it's own environment which is not mixed with other environments (there were already some issues, admittedly a minority though, which were OK with a different environment - it's due to the fact, that the script is stateful, i.e. behavior depends on the whole past seen so far).

But maybe you had something a bit different in mind and I misinterpreted it. Feel free to explain it further.

vdechef replied:
My opinion is that the most important part about testing is to ensure that the behavior of the mysql2sqlite will be correct on a whole sql file, not only on a specific query. As you said there can be interactions between statements, and if they break the resulting sqlite file, the tests should definitly catch them. In fact, it is some kind of end-to-end tests rather than unit tests.

Furthermore, I think that it would be easier to write queries in an sql file, rather than in a shell script.

I also think that this way we could easily import the sql file into a real database, and then export with mysqldump it to generate different flavors of the sql script (for example, for Windows, or a specific Linux distrib, or a given version of MySQL), without having to rewrite the whole bunch of assert : we would just have to iterate over the list of sql files and ensure that the generated sqlite file behaves the same.

Regarding the update of the sql files, it should only be done when adding a new test case. It is not meant to be modified each time you change something. Actually it would not change more often than the statements you wrote in the shell script : it is just another place to define the sql statements.

vdechef commented:
So I just changed the unit tests to match what you wanted.

I also took one of the examples you had written, just to be sure that it would work with the helpers functions I added. And thanks to this I realized that the BIT type default values needed hexstring conversion. So I wrote a function for this and it seems like it is fixed now.

vdechef commented:
While testing [a Node.js replacement for mysqldump](https://github.com/bradzacher/mysqldump), I realized that the regex used to handle `VALUES(...)` statement could easily be improved to be compatible with this library. So I made a small patch.

@dumblob I know you don't have much time, but do you think this PR has a chance to be accepted someday ?

dumblob commented:
>While testing a Node.js replacement for mysqldump, I realized that the regex used to handle VALUES(...) statement could easily be improved to be compatible with this library. So I made a small patch.

It's fine, the difference is negligible.

>@dumblob I know you don't have much time, but do you think this PR has a chance to be accepted someday ?

I'm sorry I'm really out of time now, but skimming the PR it looks very good - it's a lot of hard work and I'm glad you devoted so much time and effort to this.

One minor nit (I repeat I didn't read the changes thoroughly, so there might be other things once I'll find some more time) would be my preference to not "rename" utilities (with "rename" I'm referring to storing their names in variables and subsequently using these variables). Any good reasons for that? If we really needed to "shadow" a real utility (e.g. to increase cross-platform support or speed or whatever), then I'd definitely rather introduce a differently named function as a wrapper with an easy interface (e.g. fixed number of args) to make the need explicit.

vdechef commented:
> To answer your question - yes, I definitely want to merge this PR!

Nice, I will continue to improve it when needed.

> Any good reasons for that? If we really needed to "shadow" a real utility (e.g. to increase cross-platform support or speed or whatever), then I'd definitely rather introduce a differently named function as a wrapper with an easy interface (e.g. fixed number of args) to make the need explicit.

I totally agree. I just tried to mimic the style of existing code, but I didn't feel comfortable with this because the args of those utilities prevent their replacement without code refactoring.  
I just updated the code to remove those declarations.

vdechef commented:
Hi @dumblob,
More than 3 years later, I rebased my PR and updated it with a CI file that will run the unit tests on Pull-Requests.
Feel free to review/merge this.

vdechef commented:
I added a few more test cases. 
Thanks to this I found that:
1. hex integers longer than 16 characters are not truncated anymore (an error is raised when creating the database)
2. bit fields set with integer values are not correctly parsed (we get ASCII instead of int)

Honestly I don't have enough energy to investigate this

vdechef commented:
I just realized that the previous points were invalid:
1. by digging in our old discussions, it appear that we stopped truncating hex integers, because this causes dataloss
2. bit-fields are parsed correctly as long as one does not omit the `--hex-blob` option during mysqldump operation, as stated in the README

So I added a few more tests and pushed the code.

For the remaining test cases, I don't understand what they are supposed to test so I cannot implement them.

dumblob commented:
@vdechef this is wonderful! Many thanks! I am indeed much less active during the last years but I really appreciate all the effort you have put into this.

Now there we are considering moving everything under the organization https://github.com/mysql2sqlite - feel free to express yourself if you find this idea valuable and whether you would wish to become a member (for reviews etc.).
